src/tests/reactA11yAccessibleHeadingsRuleTests.ts

```diff
@@ -0,0 +1,281 @@
+import { Utils } from '../utils/Utils';
+import { TestHelper } from './TestHelper';
+
+describe('reactA11yAccessibleHeadingsRule', (): void => {
+    const ruleName: string = 'react-a11y-accessible-headings';
+    describe('react accessible heading rules for plain heading elements', (): void => {
+        it('should pass on when heading is non empty expression', (): void => {
+            const script: string = `
```

Comment:
It looks like the rule's logic is still running under the assumption that a single file contains all the `<h*>` s a particular page might have... but that's often not so. Try adding this test script:

```tsx
export const createBasicH1 = () => <h1>Hello, world!</h1>;
export const createBasicH2 = () => <h2>Hello!</h2>;
export const createTextH1 = (text: string) => <h1>{text}</h1>;
export const createTextPropsH1 = (text: { text: string }) => <h1>{text}</h1>;
export const createTextPropsH2 = (text: { text: string }) => <h2>{text}</h2>;
export const andAnotherH2 = () => <h2>What a rude edge case</h2>;
export const andAnotherH1 = () => <h1>Eeeeeeevil!</h1>;
```

Theoretical context to that snippet: other files would import those components and use them elsewhere. That snippet and the equivalent with classes extending from `React.Component` _(or something like them)_ should be included in your test suite and shouldn't flag any failures for having multiple headings of the same type.

The red flag in your implementation is that you're storing private member variables in the walker to keep track of heading information.. but you can really only keep track of those contents _within a particular component or `render()` function_.

Recommendation: make the rule much dumber and have it only consider counts, etc. within individual components. This should fail:

```tsx
export const someComponent = () => (
    <div>
        <h1>Hello,</h1>
        <h1>world!</h1>
    </div>
);
```

...but this shouldn't:

```tsx
export const createFirstH1 = () => <h1>Hello,</h1>;
export const createSecondH1 = () => <h1>world!</h1>;
```

Does that seem reasonable?

Reply:
@JoshuaKGoldberg sounds very reasonable and educating. 
Just as a quick update - The test snippet which you've suggested actually passes.
 
So, do you think that maybe the best approach is to remove all member variables, keep using `VisitVariableDeclaration` as the entry point, and keep implementation inside?